### PR TITLE
feature/kafka

### PR DESCRIPTION
**Pull request by DmitryOnopriienko**
@HNDaG

### CODE REVIEW

**Comment by HNDaG** [comment]
LGTM

File: internal-api/src/main/kotlin/com/ajaxproject/api/internal/warehousesvc/KafkaTopic.kt [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+import com.ajaxproject.api.internal.warehousesvc.NatsSubject.WAREHOUSE_PREFIX
+
+object KafkaTopic {
+
+    object Waybill {
+
+        private const val WAYBILL_PREFIX = "${WAREHOUSE_PREFIX}.waybill"
+
+        const val CREATION = "$WAYBILL_PREFIX.creation"
```
Comment (ajax-gnatiuk-s) [inline internal-api/src/main/kotlin/com/ajaxproject/api/internal/warehousesvc/KafkaTopic.kt]:
```suggestion
        const val CREATED = "$WAYBILL_PREFIX.created"
```

Kafka event - some event has happened. So, use verb 3rd form

File: internal-api/src/main/kotlin/com/ajaxproject/api/internal/warehousesvc/KafkaTopic.kt [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package com.ajaxproject.api.internal.warehousesvc
+
+import com.ajaxproject.api.internal.warehousesvc.NatsSubject.WAREHOUSE_PREFIX
```
Comment (ajax-gnatiuk-s) [inline internal-api/src/main/kotlin/com/ajaxproject/api/internal/warehousesvc/KafkaTopic.kt]: It's better to have independent prefixes for NATS and Kafka.

It's ok to have two constants with the same value.

File: wm-system/src/main/kotlin/com/ajaxproject/warehouse/service/WaybillServiceImpl.kt [hunk 3]
```diff
@@ -123,6 +132,22 @@ class WaybillServiceImpl(
...
+                date = Timestamp.newBuilder()
+                    .setSeconds(
+                        this@mapToCreatedEventProto.date
+                            .atStartOfDay()
+                            .atOffset(ZoneOffset.UTC)
+                            .toEpochSecond()
+                    )
+                    .build()
+            }.build()
+        }.build()
```
Comment (ajax-gnatiuk-s) [inline wm-system/src/main/kotlin/com/ajaxproject/warehouse/service/WaybillServiceImpl.kt]:
```suggestion
        WaybillCreatedEvent.newBuilder().also { event ->
            event.waybillBuilder.also { waybill ->
                waybill.id = id
                waybill.customerId = customer.id
                waybill.dateBuilder.seconds = date.atStartOfDay().atOffset(ZoneOffset.UTC).toEpochSecond()
            }
        }.build()
```

Looks better, dosn't it?

File: wm-system/src/main/kotlin/com/ajaxproject/warehouse/service/WaybillServiceImpl.kt [hunk 4]
```diff
@@ -47,6 +52,10 @@ class WaybillServiceImpl(
                 findValidEntitiesOrError(productIds)
             }
             .then(createWaybillAndMapToDto(createDto))
+            .flatMap {
+                kafkaProducerService.sendWaybillCreationEvent(it.mapToCreatedEventProto())
+                it.toMono()
```
Comment (ajax-gnatiuk-s) [inline wm-system/src/main/kotlin/com/ajaxproject/warehouse/service/WaybillServiceImpl.kt]:
```suggestion
            .flatMap { waybillDto ->
                kafkaProducerService.sendWaybillCreationEvent(waybillDto.mapToCreatedEventProto())
                    .thenReturn(waybillDto)
```

File: internal-api/src/main/kotlin/com/ajaxproject/api/internal/warehousesvc/KafkaTopic.kt (hunk 1 above)
Comment (DmitryOnopriienko) [reply internal-api/src/main/kotlin/com/ajaxproject/api/internal/warehousesvc/KafkaTopic.kt]: Fixed

File: internal-api/src/main/kotlin/com/ajaxproject/api/internal/warehousesvc/KafkaTopic.kt (hunk 2 above)
Comment (DmitryOnopriienko) [reply internal-api/src/main/kotlin/com/ajaxproject/api/internal/warehousesvc/KafkaTopic.kt]: Done

File: wm-system/src/main/kotlin/com/ajaxproject/warehouse/service/WaybillServiceImpl.kt (hunk 3 above)
Comment (DmitryOnopriienko) [reply wm-system/src/main/kotlin/com/ajaxproject/warehouse/service/WaybillServiceImpl.kt]: Done

File: wm-system/src/main/kotlin/com/ajaxproject/warehouse/service/WaybillServiceImpl.kt (hunk 4 above)
Comment (DmitryOnopriienko) [reply wm-system/src/main/kotlin/com/ajaxproject/warehouse/service/WaybillServiceImpl.kt]: Refactored

**Review by ajax-gnatiuk-s — APPROVED** [review]

**Pull request merged**